### PR TITLE
fix(RouterStore): Throw RouterState should trigger beforeExit, beforeEnter, onExit, onEnter

### LINKED ISSUE #42: Throw RouterState won't trigger onEnter
```ts
const routes: Routes[] = [
    { name:"x"    , pattern:"/x/:id",
        onEnter: async (fromState, toState, routerStore) => {
            // ...
            throw (new RouterState('y', {id:"0"}));
        }},
    { name:"y"    , pattern:"/y/:id",
        onEnter: async (fromState, toState, routerStore) => {
            // won't be trigger if redirect from state x
        }},
]
```

When `onEnter` of state `x` `throw (new RouterState('y'))`,
it redirect to state `y` but won't trigger `onEnter` of state `y`.

### PR DESCRIPTION
**Pull request by rexpan**
```ts
const routes: Routes[] = [
    { name:"x"    , pattern:"/x/:id",
        onEnter: async (fromState, toState, routerStore) => {
            // ...
            throw (new RouterState('y', {id:"0"}));
        }},
    { name:"y"    , pattern:"/y/:id",
        onEnter: async (fromState, toState, routerStore) => {
            // won't be trigger if redirect from state x
        }},
]
```

When `onEnter` of state `x` `throw (new RouterState('y'))`,
it redirect to state `y` should trigger `onEnter` of state `y`.

fix #42

### CODE REVIEW

File: src/router-store.ts [hunk 1]
```diff
@@ -158,7 +158,7 @@ export class RouterStore {
                     'states are equal, skipping'
                 );
             }
-            return Promise.resolve(toState);
+            return toState;
```
[the remark this answers is not in the document]
Comment (rexpan) [reply src/router-store.ts]: @nareshbhatia I changed `transition` to `async` so we can return value and `asycn` will wrap it with Promise.

File: src/router-store.ts [hunk 2]
```diff
@@ -176,31 +176,22 @@ export class RouterStore {
...
+            if (beforeEnter) await beforeEnter(fromState, toState, this);
+            if (onExit) await onExit(fromState, toState, this);
+            if (onEnter) await onEnter(fromState, toState, this);
+            this.setRouterState(toState);
+            return toState;
+        } catch (redirectState) {
+            if (redirectState instanceof RouterState === false) {
+                throw new Error('toState is undefined');
+            }
+            this.goTo(redirectState);
```
[the remark this answers is not in the document]
Comment (rexpan) [reply src/router-store.ts]: @nareshbhatia 

I updated the PR for fixing the test.

---

> There are 3 lint errors in this PR around this area.

I run `lint` and not get any errors.
If you still encounter and errors, could you send me the output.

---

> Can you provide a concrete use case where you need to call a hook function when you hit an error?

I expect when enter route `y`, `hook` is called so I can fetch data, update store, set up state, ...
But when I redirected from route `x`, no `hook` function is called.
For workaround, I have to duplicate the code for setting up before `throw` to state `y`.

**Comment by nareshbhatia** [comment]
@rexpan, thanks for making requested changes.

**Pull request merged**